### PR TITLE
Charm v2 metadata with Systems support

### PR DESCRIPTION
**Pull request by hpidcock**
This PR adds Charm v2 metadata with Systems support

- Systems support with Channels and Resources.
- Backwards compatability with Series for well-known Systems.
- Container support.
- Architecture and Platform support.

### ComputedSeries

ComputedSeries is a newly introduced charm metadata method for v2 charms that use systems.
Since there is a large amount of logic in Juju that depends on series as a string, ComputedSeries returns systems as either their canonical series name like "focal" or as a complex string "system#resource=charm-resource-name".

Systems as a series string can support resource images or they can support simple os and unknown channels like "system#os=ubuntu#channel=21.04/stable"

### CODE REVIEW

**Review by ycliuhw — APPROVED** [review]
lgtm thanks

File: meta.go [hunk 1]
```diff
@@ -273,8 +276,46 @@ type Meta struct {
...
+)
+
+type Architecture string
+
+const (
+	AMD64   Architecture = "amd64"
+	ARM64   Architecture = "arm64"
+	PPC64EL Architecture = "ppc64el"
+	S390X   Architecture = "s390x"
+)
```
Comment (wallyworld) [inline meta.go:300]: We have arch stuff under juju/utils.
Ideally one day we'd move it to juju/arch and get this from there

File: meta.go [hunk 2]
```diff
@@ -273,8 +276,46 @@ type Meta struct {
...
 	Terms          []string                 `bson:"terms,omitempty" json:"Terms,omitempty"`
 	MinJujuVersion version.Number           `bson:"min-juju-version,omitempty" json:"min-juju-version,omitempty"`
+
+	Systems       []systems.System     `bson:"systems,omitempty" json:"systems,omitempty" yaml:"systems,omitempty"`
+	Platforms     []Platform           `bson:"platforms,omitempty" json:"platforms,omitempty" yaml:"platforms,omitempty"`
+	Architectures []Architecture       `bson:"architectures,omitempty" json:"architectures,omitempty" yaml:"architectures,omitempty"`
+	Containers    map[string]Container `bson:"containers,omitempty" json:"containers,omitempty" yaml:"containers,omitempty"`
 }
 
+type Platform string
```
Comment (wallyworld) [inline meta.go:286]: Since this is a library, we should be good at commenting public artefacts

File: meta.go [hunk 3]
```diff
@@ -523,11 +593,28 @@ func parseMeta(m map[string]interface{}) (*Meta, error) {
...
-	resources, err := parseMetaResources(m["resources"])
+	meta.Resources, err = parseMetaResources(m["resources"])
+	if err != nil {
+		return nil, err
+	}
+
+	// v2 parsing
+	meta.Systems, err = parseSystems(m["systems"], meta.Resources, true)
+	if err != nil {
+		return nil, err
```
Comment (wallyworld) [inline meta.go:604]: maybe annotate these to give a clue what top level key failed to parse?

File: meta.go [hunk 4]
```diff
@@ -999,6 +1138,126 @@ func parseDeployment(deployment interface{}, charmSeries []string, storage map[s
...
+	for _, v := range input.([]interface{}) {
+		system := systems.System{}
+		systemMap := v.(map[string]interface{})
+		if value, ok := systemMap["os"]; ok {
+			system.OS = value.(string)
+		}
+		if value, ok := systemMap["channel"]; ok {
+			system.Channel, err = channel.Parse(value.(string))
+			if err != nil {
+				return nil, errors.Trace(err)
```
Comment (wallyworld) [inline meta.go:1156]: same here, worth annotating

File: meta.go [hunk 5]
```diff
@@ -999,6 +1138,126 @@ func parseDeployment(deployment interface{}, charmSeries []string, storage map[s
...
+		}
+		if value, ok := mountMap["location"].(string); ok {
+			mount.Location = value
+		}
+		if mount.Storage == "" {
+			return nil, errors.Errorf("storage must be specifed on mount")
+		}
+		if mount.Location == "" {
+			return nil, errors.Errorf("location must be specifed on mount")
+		}
```
Comment (wallyworld) [inline meta.go:1252]: no tests for these that i can see

File: meta.go [hunk 6]
```diff
@@ -999,6 +1138,126 @@ func parseDeployment(deployment interface{}, charmSeries []string, storage map[s
...
+	return architectures, nil
+}
+
+func parseContainers(input interface{}, resources map[string]resource.Meta, platforms []Platform, storage map[string]Storage) (map[string]Container, error) {
+	var err error
+	if input == nil {
+		return nil, nil
+	}
+	if len(platforms) != 1 || platforms[0] != PlatformKubernetes {
+		return nil, errors.Errorf("containers are currently only supported on kubernetes platform")
```
Comment (wallyworld) [inline meta.go:1211]: test?

File: meta.go [hunk 7]
```diff
@@ -999,6 +1138,126 @@ func parseDeployment(deployment interface{}, charmSeries []string, storage map[s
...
+		}
+		if system.Resource != "" {
+			if r, ok := resources[system.Resource]; !ok {
+				return nil, errors.NotFoundf("referenced resource %q", system.Resource)
+			} else if r.Type != resource.TypeContainerImage {
+				return nil, errors.Errorf("referenced resource %q is not a %s",
+					system.Resource,
+					resource.TypeContainerImage.String())
+			}
+		}
```
Comment (wallyworld) [inline meta.go:1177]: add tests

**Review by wallyworld — APPROVED** [review]

File: meta.go (hunk 7 above)
Comment (hpidcock) [reply meta.go:1187]: added 7 more tests.

File: meta.go (hunk 1 above)
Comment (hpidcock) [reply meta.go:306]: Lets come back to this when architecture support is added to the rest of juju

**Comment by hpidcock** [comment]
$$merge$$

**Pull request merged**